processing/src/main/java/org/apache/druid/error/DruidException.java

```diff
@@ -0,0 +1,389 @@
+        this,
+        errorCode,
+        targetPersona,
+        category,
+        StringUtils.format("%s: %s", StringUtils.nonStrictFormat(msg, args), getMessage())
+    ).withContext(context);
+  }
+
+  /**
+   * The persona that the message on a DruidException is targetting
+   */
+  public enum Persona
+  {
+    /**
+     * Represents the end-user, a persona who is issuing queries to the Druid Query APIs
+     */
+    USER,
+    /**
+     * Represents an administrative user, a persona who is interacting with admin APIs and understands Druid query
+     * concepts without necessarily owning the infrastructure and operations of the cluster
+     */
+    ADMIN,
+    /**
+     * Represents a persona who actively owns and operates the cluster.  This persona is not assumed to understand
+     * Druid query concepts, but instead understand cluster operational concepts.
+     */
+    OPERATOR,
+    /**
+     * Represents someone who has all of the context and knowledge to be actively diving into the Druid codebase.
+     * This persona exists as a catch-all for anything that is so deep and technically in the weeds that it is not
+     * possible to make a message that will make sense to a different persona.  Generally speaking, there is a hope
+     * that only DEFENSIVE error messages will target this persona.
+     */
+    DEVELOPER
+  }
+
+  /**
+   * Category of error.  The simplest way to describe this is that it exists as a classification of errors that
+   * enables us to identify the expected response code (e.g. HTTP status code) of a specific DruidException
+   */
+  public enum Category
+  {
+    /**
+     * Means that the exception is being created defensively, because we want to validate something but expect that
+     * it should never actually be hit.  Using this category is good to provide an indication to future reviewers and
+     * developers that the case being checked is not intended to actually be able to occur in the wild.
+     */
+    DEFENSIVE(500),
+    /**
+     * Means that the input provided was malformed in some way.  Generally speaking, it is hoped that errors of this
+     * category have messages written either targetting the USER or ADMIN personas as those are the general users
+     * of the APIs who could generate invalid inputs.
+     */
+    INVALID_INPUT(400),
+    /**
+     * Means that the error is a problem with authorization.
+     */
+    UNAUTHORIZED(401),
+    /**
+     * Means that some capacity limit was exceeded, this could be due to throttling or due to some system limit
+     */
+    CAPACITY_EXCEEDED(429),
+    /**
+     * Means that the query was canceled for some reason
+     */
+    CANCELED(500),
+    /**
+     * Indicates a server-side failure of some sort at runtime
+     */
+    RUNTIME_FAILURE(500),
+    /**
+     * A timeout happened
+     */
+    TIMEOUT(504),
+    /**
+     * Indicates some unsupported behavior was requested.  TODO
+     */
+    UNSUPPORTED(501),
```

Comment:
so lets say someone issues a `UNION` query with de-duplicating while only supports `UNION ALL`. would you consider that as a `501` or `400`? Such planning errors are very common. I am concerned about the error code for such scenarios because a cluster admin might have set an alert on 5xx but not on 4xx. Since the former will indicate the server being bad while the latter indicates the user being bad.

Reply:
TBH, I'm fine with both.  The 501 code is actually just a hold-over from the QueryException rather than something I'm actually introducing.  If anything, reviewing this change has caused it to be more clear that we are returning 501s in some instances, leading to a question of "WTF", why?  Which I think is a totally valid question ;).

Fwiw, I have a TODO in there because this is something I wanted to revisit myself.  At this point, the only usages of this in the code in this PR are

https://github.com/apache/druid/pull/14004/files#diff-483e089932b8faa5ca9669607eab7fcba883d27e2899a7c9fba2f37afb9de97bR599-R602

Which is what happens when we get a query planning failure and no message at all.  I don't know if/when this actually happens, so I'm unclear on whether it's actually meaningful.  That said, with the switch to the `DruidException`, we should be able to eliminate the whole `plannerContext().getPlanningError()` thing that is the source of potentially not having a message (and also is the source of red-herring error messages more often than not), which should, in turn, eliminate the need for this category.  Regardless, it needs to exist to maintain what QueryException is doing...